### ISSUE TITLE
Support storing, querying on arbitrary properties for Survey

### ISSUE THREAD

**Issue by rsutphin** [NUARIG/surveyor#277]
An applications may contain more than one survey, or multiple versions of the same survey. Such applications may need to select surveys based on or report about surveys taken using more information than is allowed for in the Survey model. Some examples for NCS instruments include the version of the instrument definition on which the survey is based or the source instrument's type code.

Such applications have the option of defining additional attributes on the Survey model. Then they face another question: how to populate these additional attributes? If they are related to how the survey is defined (as with the NCS examples), it would make sense to keep them with the corresponding survey. Unfortunately the survey DSL is not extensible, so this is not an option.

While making the survey DSL extensible might be a good idea later, for now I propose adding a set of arbitrary properties to the Survey model. These could be defined in the DSL:

``` ruby
survey 'INS_BIO_Sample' do
  property :instrument_version, '2.0'
  property :instrument_type_code, '7'

  section 'Start' do
    # ...
  end
end
```

Then you could read them from a survey instance:

``` ruby
s = Survey.first
s.properties[:instrument_version] # => "2.0"
```

Or query for matching surveys:

``` ruby
Survey.find_by_property(:instrument_version, '2.0').collect(&:title) # => ['INS_BIO_Sample']
```

**Comment by rsutphin** [comment]
> Then you could read them from a survey instance:

Or, perhaps easier to implement:

``` ruby
s = Survey.first
s.property(:instrument_version) # => "2.0"
s.properties # => { :instrument_version => "2.0", :instrument_type_code => '7' }; read-only
```

**Comment by yoon** [comment]
Sounds great to me, although I don't think rails would support Survey.find_by_property by default.

I'm sure we could write methods, though, that would work around this. Here's one solution, although it uses serialized text in a single field rather than a to_many model

https://github.com/moiristo/dynamic_attributes (serialized)

**Comment by rsutphin** [comment]
> I don't think rails would support Survey.find_by_property by default.

Yeah, I'm sure it wouldn't. I just wanted to sketch out what a desirable API would be for discussion purposes.

> Here's one solution [...]

I'd be inclined to do an associated model (the relational term for this would be an entity-attribute-value table, or EAV) since we would want to be able to support efficient querying. I only found one EAV library and it doesn't seem to be actively maintained, but it's also not something that's hard to do by hand with AR.

**Comment by yoon** [comment]
Great. I had trouble finding a good EAV library as well.

**Comment by yoon** [comment]
@rsutphin https://github.com/200proof/eav_hashes seems to best match your desirable API the closest

Similar projects include:
https://github.com/kostyantyn/hydra_attribute
https://github.com/coffeeaddict/has_eav/commits/master
https://github.com/rpheath/acts_as_configurable/commits/master  (stale)
https://github.com/hypomodern/flex-attributes/commits/master (stale)

**Comment by rsutphin** [comment]
I agree — 200proof/eav_hashes looks great API-wise.
